### PR TITLE
1232 Hierarchical project list for org homepage

### LINKED ISSUE #1232: Hierarchical project list for org homepage
Organizations like the Library of Virginia may have several projects (WW1 questionnaires, pension applications, photograph collections) uploaded as individual collections, with document sets under each collection used to organize material by county.  They'd like to be able to present these on their organization landing page as a tree:
```
WW1 Questionnaires
  County A
  County B
  County C
1812 Pension Applications
  County A
  County B
  County C
```

Since we treat document sets as if they were top-level collections (which is valid for other use cases) they are instead organized in a flat, alphabetical format:

```
1812 Pension Applications
County A
County A
County B
County B
County C
County C
WW1 Questionnaires
```

This layout is confusing for volunteers on each project.

We'd like to re-model the organization home page to present a new hierarchical display *if a project has public collections with document sets*.  So the logic should be something along the lines of:

```
for each collection-or-set
  if collection-or-set.is_a Collection && collection.public?
    print collection title/description/link (as top level)
    for each collection.document_set
      print document set title/link (as member)
  if collection-or-set.is_a DocumentSet && set.collection.private?
  print document set title/link (as top-level)
```

### PR DESCRIPTION
**Pull request by lortza**
Resolves #1232 [Hierarchical project list for org homepage](https://waffle.io/benwbrum/fromthepage/cards/5bdb16fe56fc50003b381cdc)

## Problem
For some conditional cases, displaying document sets and collections as visual equals is confusing:
```
1812 Pension Applications
County A
County A
County B
County B
County C
County C
WW1 Questionnaires
```
In these cases, we'd prefer to have a hierarchy:
```
WW1 Questionnaires
  County A
  County B
  County C
1812 Pension Applications
  County A
  County B
  County C
```

### Before
Alphabetical order and collections & doc sets as equals
screenshot 2018-11-05 13 24 44

### After
Alphabetical order, but with doc sets nested under their collection title for public collections.
Example without intro text
screenshot 2018-11-05 13 24 20

Example with intro text
screenshot 2018-11-05 13 33 34

Collection with private doc sets does not show doc sets
screenshot 2018-11-05 12 07 05

## While Doing this Work...
I moved some instance variables into the local variable space in order to reduce any unexpected or unintended reach. You'll see some notes about this on individual files.

### CODE REVIEW

File: app/controllers/dashboard_controller.rb [hunk 1]
```diff
@@ -47,8 +47,8 @@ def index
 
   def collections_list
     collections = Collection.includes(:owner).distinct
-    @document_sets = DocumentSet.includes(:owner).distinct
-    @collections = (collections + @document_sets).sort { |a,b| a.title <=> b.title }
+    document_sets = DocumentSet.includes(:owner).distinct
+    @collections_and_document_sets = (collections + document_sets).sort { |a,b| a.title <=> b.title }
```
Comment (lortza) [inline app/controllers/dashboard_controller.rb]: I renamed this variable to make it more clear what kind of data it is housing.

File: app/views/dashboard/_alphabetical_collections_and_document_sets.html.slim [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+          .collection
+            -unless cds.picture.blank?
+              .collection_image
+                =image_tag(cds.picture_url(:thumb), alt: cds.title)
+            .collection_details
+              h4.collection_title =link_to cds.title, collection_path(cds.owner, cds)
+              -unless snippet.empty?
+                p.collection_snippet =snippet
+              .collection_summary
+                span =="Owned by #{owner_link}"
```
Comment (lortza) [inline app/views/dashboard/_alphabetical_collections_and_document_sets.html.slim]: Though the diff makes this look like a new file, it is just a rename
- from: `app/views/dashboard/_collection_index.html.slim`
- to: `app/views/dashboard/_alphabetical_collections_and_document_sets.html.slim`

File: app/views/user/_owner_profile.html.slim [hunk 3]
```diff
@@ -39,11 +39,15 @@ section.user-profile
...
             
 .columns.list
-  article.maincol
-    =render '/dashboard/collection_index'
+  article.maincol  
+    -if any_public_collections_with_document_sets?(collections_and_document_sets)
+      =render partial: '/dashboard/hierarchical_collections_and_document_sets', locals: {collections_and_document_sets: collections_and_document_sets}
+    -else
+      =render partial: '/dashboard/alphabetical_collections_and_document_sets', locals: {collections_and_document_sets: collections_and_document_sets}
+
```
Comment (lortza) [inline app/views/user/_owner_profile.html.slim]: This is the main logic that decides if we use the hierarchical or alphabetical look.

File: app/views/user/_owner_profile.html.slim [hunk 4]
```diff
@@ -1,35 +1,35 @@
--user_name = @user.display_name.presence || @user.login
+-user_name = user.display_name.presence || user.login
```
Comment (lortza) [inline app/views/user/_owner_profile.html.slim]: This line and similar ones are just a change from using `@` instance variables to local ones in order to reduce unexpected reach. You'll see them being passed explicitly to this partial in  `app/views/user/profile.html.slim`.

**Comment by lortza** [comment]
I'm getting a clean run of tests locally. Any ideas about why the translation-related tests are failing on some of the servers?

**Comment by lortza** [comment]
I don't feel 100% on this logic and would love to review it with someone who has a stronger sense of needs to go where.

File: app/controllers/dashboard_controller.rb (hunk 1 above)
Comment (benwbrum) [reply app/controllers/dashboard_controller.rb]: Hooray!

File: app/views/dashboard/_alphabetical_collections_and_document_sets.html.slim (hunk 2 above)
Comment (benwbrum) [reply app/views/dashboard/_alphabetical_collections_and_document_sets.html.slim]: Looks good.

**Pull request merged**